### PR TITLE
Fix ColorPicker deferred mode not working for sliders.

### LINKED ISSUE #80915: Color picker sliders modification will call color changed signal even when deferred mode is active
### Godot version

v4.2.dev.custom_build [0511f9d9a]

### System information

macOS 13.3.0 - Vulkan (Forward+) - integrated Apple M1 Pro - Apple M1 Pro (8 Threads)

### Issue description

Even when the deferred mode from Color picker is enabled modifying the value from one of the sliders will triger the `color_changed` signal for every little change which will cause performance issue in some cases.
This does not happen with the main widget only with the sliders. 

### Steps to reproduce

1. Create a scene with a color picker and set deferred 
2. connect the signal to any high consuming function that you should not call too often (a lot of prints should be good eenough)
3. run the scene and try to modify the value through the sliders

### Minimal reproduction project

[Test.zip](https://github.com/godotengine/godot/files/12415304/Test.zip)

### PR DESCRIPTION
**Pull request by ajreckof**
Fixes #80915

### CODE REVIEW

File: scene/gui/color_picker.h [hunk 1]
```diff
@@ -248,7 +248,7 @@ class ColorPicker : public VBoxContainer {
 	void create_slider(GridContainer *gc, int idx);
 	void _reset_sliders_theme();
 	void _html_submitted(const String &p_html);
-	void _value_changed(double);
+	void _value_changed(bool force_signal);
```
Comment (dalexeev) [inline scene/gui/color_picker.h:251]:
```suggestion
	void _value_changed(bool p_force_signal);
```

**Comment by KoBeWi** [comment]
Clicking the slider does not emit changed signal in deferred mode. You need to drag it at least 1 px. SpinBox is also broken (note that you can drag SpinBox too).

**Comment by ajreckof** [comment]
> It's fine now, but there is a little inconsistency. Clicking the slider for the first time emits the signal immediately, while it doesn't happen when clicking the color box/circle. This is still an improvement, so I think that small inconsistency is ok (unless you can fix it easily?). 

I found it just after pushing I was looking at it. I'm not 100% sure why it is happening but should be pretty easy to fix 

>Dragging SpinBox still emits signal continuously. SpinBox doesn't expose any drag signals though, so it can be fixed later.

Yeah that was what I was going to say the class feels a bit lacking :(

**Comment by ajreckof** [comment]
So I fixed the slider trigerring `color_changed` signal at the start of dragging, and at the same time it made me realise and fix that when clicking the slider it would trigger the signal twice.

**Comment by KoBeWi** [comment]
SpinBoxes are broken in deferred mode again 🙃

**Comment by KoBeWi** [comment]
What are the Range changes for?

File: scene/gui/color_picker.h [hunk 2]
```diff
@@ -248,7 +249,9 @@ class ColorPicker : public VBoxContainer {
 	void create_slider(GridContainer *gc, int idx);
 	void _reset_sliders_theme();
 	void _html_submitted(const String &p_html);
-	void _value_changed(double);
+	void _drag_started();
+	void _value_changed();
+	void _drag_ended();
```
Comment (KoBeWi) [inline scene/gui/color_picker.h:254]:
```suggestion
	void _slider_drag_started();
	void _value_changed();
	void _slider_drag_ended();
```
Makes it more clear.

File: scene/gui/color_picker.cpp [hunk 3]
```diff
@@ -383,8 +387,17 @@ void ColorPicker::_value_changed(double) {
 		last_color = color;
 	}
 
-	_set_pick_color(color, false);
-	emit_signal(SNAME("color_changed"), color);
+	set_pick_color(color);
```
Comment (KoBeWi) [inline scene/gui/color_picker.cpp:390]: Why `_set_pick_color(color, false);` was changed? This is a callback for slider value changed, so updating sliders again is redundant.

**Review by KoBeWi — APPROVED** [review]
Works correctly now, left a few comments.

The Range changes make `_drag_started()` emitted before value changed. *Probably* it's fine though.

**Comment by ajreckof** [comment]
I added both of your recommendation
I'm not 100% confident on the Range change but it is needed. I'll try to be as clear as possible on the reason that lead me to this modification. 
So first since the `SpinBox` and the `Slider` share the same range whenever you modify one the other is updated but also both emit the `value_changed` signal. This means when you receive a `value_changed` signal from one you don't wether it is from one or the other. 
What we want to achieve is that in deferred mode only at the drag end `color_changed` so the easiest idea is to just prevent the emission of the signal when the `value_changed` signal is received and emit the signal when the `drag_ended` signal is received. Unfortunately because of what I said earlier this isn't possible because to prevent the emission of the signal when the `value_changed` is received you can't differentiate between it being emitted from one or the other, therefore you will just never emit `color_changed` when changing value through the `SpinBox`. 
So the other way is to detect when you are effectively doing a drag which means between the reception of `drag_started` and the reception of `drag_ended`. This works well but unfortunately this has a secondary problem which is the first `value_changed` signal is emitted before the `drag_started` signal which in trns mean you won't be able to know it is from the drag. So I reordered them but as to reduce the change I made sure that only their order were changed. Which means I had to update the value of the slider first without emitting the `value_changed` signal and then first emitting `drag_started` and then emit the `value_changed` signal.

If SpinBox ever gets `drag_started`/`drag_ended` signals just connecting them to the same function will fix the drag for SpinBox too.

File: scene/gui/color_picker.h (hunk 2 above)
Comment (ajreckof) [reply scene/gui/color_picker.h]: I'm not 100% sure about `_value_changed` change. I feel like `_slider_value_changed` would hide the fact that it is also called when SpinBox is used. Maybe `_slider_or_spin_value_changed`?

Comment (KoBeWi) [reply scene/gui/color_picker.h]: In ColorPicker, SpinBoxes are treated as part of Sliders, so the name would be correct as it refers to the picker's "slider".

Comment (ajreckof) [reply scene/gui/color_picker.h]: 😲😲😲

File: scene/gui/slider.cpp [hunk 4]
```diff
@@ -68,14 +68,15 @@ void Slider::gui_input(const Ref<InputEvent> &p_event) {
...
 				} else {
-					set_as_ratio(((double)grab.pos - (grab_width / 2.0)) / max);
+					set_as_ratio_no_signal(((double)grab.pos - (grab_width / 2.0)) / max);
 				}
 				grab.active = true;
 				grab.uvalue = get_as_ratio();
-
 				emit_signal(SNAME("drag_started"));
+				_notify_shared_value_changed();
+
```
Comment (akien-mga) [inline scene/gui/slider.cpp:79]: Unnecessary empty line.

**Review by akien-mga — COMMENTED** [review]
I see this adds more `_no_signal` helpers. We should aim to resolve the discussion on #76432 before adding more methods like this (though this one isn't exposed so it's not too critical).

File: scene/gui/slider.cpp [hunk 5]
```diff
@@ -68,14 +68,15 @@ void Slider::gui_input(const Ref<InputEvent> &p_event) {
 				double grab_height = (double)grabber->get_height();
 				double max = orientation == VERTICAL ? get_size().height - grab_height : get_size().width - grab_width;
 				if (orientation == VERTICAL) {
-					set_as_ratio(1 - (((double)grab.pos - (grab_height / 2.0)) / max));
+					set_as_ratio_no_signal(1 - (((double)grab.pos - (grab_height / 2.0)) / max));
```
Comment (KoBeWi) [inline scene/gui/slider.cpp:71]: The method is internal, so I paid no attention, but technically you can do
```suggestion
					set_block_signals(true);
					set_as_ratio(1 - (((double)grab.pos - (grab_height / 2.0)) / max));
					set_block_signals(false);
```

File: scene/gui/range.h [hunk 6]
```diff
@@ -79,6 +81,7 @@ class Range : public Control {
 	void set_step(double p_step);
 	void set_page(double p_page);
 	void set_as_ratio(double p_value);
+	void set_as_ratio_no_signal(double p_value);
```
Comment (akien-mga) [inline scene/gui/range.h:84]: This is no longer added, so should be removed.

Comment (akien-mga) [reply scene/gui/range.h:84]: Bump @ajreckof

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**